client/state-db/src/pruning.rs

```diff
@@ -24,418 +24,964 @@
+		(self.queue.len() - self.pending_prunings) as u64
 	}
 
-	pub fn next_hash(&self) -> Option<BlockHash> {
-		self.death_rows.get(self.pending_prunings).map(|r| r.hash.clone())
+	pub fn next_hash(&mut self) -> Result<Option<BlockHash>, Error<D::Error>> {
+		self.queue.get_hash(self.base, self.pending_prunings)
 	}
 
 	pub fn mem_used(&self) -> usize {
 		0
 	}
 
+	// Return the block number of the first block that not been pending pruned
 	pub fn pending(&self) -> u64 {
-		self.pending_number + self.pending_prunings as u64
+		self.base + self.pending_prunings as u64
 	}
 
-	pub fn have_block(&self, hash: &BlockHash) -> bool {
-		self.death_rows.iter().skip(self.pending_prunings).any(|r| r.hash == *hash)
+	fn is_empty(&self) -> bool {
+		self.queue.len() <= self.pending_prunings
+	}
+
+	// Check if a block is in the pruning window and not be pruned yet
+	pub fn have_block(&self, hash: &BlockHash, number: u64) -> HaveBlock {
+		// if the queue is empty or the block number exceed the pruning window, we definitely
+		// do not have this block
+		if self.is_empty() ||
+			!(number >= self.pending() && number < self.base + self.queue.len() as u64)
+		{
+			return HaveBlock::NotHave
+		}
+		self.queue.have_block(hash, (number - self.base) as usize)
 	}
 
 	/// Prune next block. Expects at least one block in the window. Adds changes to `commit`.
-	pub fn prune_one(&mut self, commit: &mut CommitSet<Key>) {
-		if let Some(pruned) = self.death_rows.get(self.pending_prunings) {
+	pub fn prune_one(&mut self, commit: &mut CommitSet<Key>) -> Result<(), Error<D::Error>> {
+		if let Some(pruned) = self.queue.get(self.base, self.pending_prunings)? {
 			trace!(target: "state-db", "Pruning {:?} ({} deleted)", pruned.hash, pruned.deleted.len());
-			let index = self.pending_number + self.pending_prunings as u64;
-			commit.data.deleted.extend(pruned.deleted.iter().cloned());
+			let index = self.base + self.pending_prunings as u64;
+			commit.data.deleted.extend(pruned.deleted.into_iter());
 			commit.meta.inserted.push((to_meta_key(LAST_PRUNED, &()), index.encode()));
-			commit.meta.deleted.push(pruned.journal_key.clone());
+			commit
+				.meta
+				.deleted
+				.push(to_journal_key(self.base + self.pending_prunings as u64));
 			self.pending_prunings += 1;
 		} else {
 			warn!(target: "state-db", "Trying to prune when there's nothing to prune");
 		}
+		Ok(())
 	}
 
 	/// Add a change set to the window. Creates a journal record and pushes it to `commit`
-	pub fn note_canonical(&mut self, hash: &BlockHash, commit: &mut CommitSet<Key>) {
+	pub fn note_canonical(
+		&mut self,
+		hash: &BlockHash,
+		number: u64,
+		commit: &mut CommitSet<Key>,
+	) -> Result<(), Error<D::Error>> {
+		if self.base == 0 && self.queue.len() == 0 && number > 0 {
+			// assume that parent was canonicalized
+			self.base = number;
+		} else if (self.base + self.queue.len() as u64) != number {
+			return Err(Error::StateDb(StateDbError::InvalidBlockNumber))
+		}
```

Comment:
This change is subject to:
https://github.com/paritytech/substrate/blob/055453ebdf00db9e59ef5af0c59c1b4e39267813/client/state-db/src/noncanonical.rs#L251-L258

namely that the first block put into the pruning window may not be block 0. So in the previous implementation the block number in the pruning window is incorrect, but it is okay because its public functions didn't take any block number as parameter (perhaps this is why the previous `have_block` needs to brute force search all blocks)

But this PR optimized `have_block` by using block number indexing the query block thus causing the `syncs_state` test to fail, thus we need add this change to keep consistent with the `non_canonical` layer. Also, I have reverted the change in `sc_state_db::tests::make_test_db` (previously mentioned in https://github.com/paritytech/substrate/pull/11980#discussion_r941654449) to add test cases (along with `syncs_state`) to this change.
/cc @arkpar @cheme

Reply:
@NingLin-P , not sure if related, but I tried to run polkadot patched with this PR and `prune` seems to infiniloop ("Trying to prune when there's nothing to prune").

Reply:
Do you have any log (`trace` level), I'd like to have a check

Reply:
https://gist.github.com/cheme/f1ea299959f23d94636160cb00ace674
But actually does not seems related to your pr, there is something else I need to find (this only happen with parity-db).

Reply:
Ok, I am stupid!!! rocksdb do not use the DB backend. So yes trace attached possibly shows an issue with this PR.
Got them with `./target/release/polkadot -d ./tmp1 --state-pruning 256 --log state-db=trace --log client-db=trace 2&> log`
Note that starting with a window of 900 blocks, I don't see the issue anymore.

Reply:
> actually seems more likely we are trying to get a key from db that is still in commit_set. (I see insert at some point and miss 1 millis after).

Yes, I also found the cause of this issue is db return `None` for some blocks that had previously been added to the `commit_set`.

Reply:
I was thinking fix this issue by 
1. Adding an additional cache of recent added blocks, so we can get blocks that not yet commit to the database from this cache
2. Or, don't count block that not yet commit to the database for `window_size`, so we will not prune block that not yet commit to the database

but both approach required to do some cleanup work (clear cache or adjust `window_size`) in `apply_pending/revert_pending`, and seems these functions may be called concurrently which may mix up the state: 
https://github.com/paritytech/substrate/blob/bdd139989cb4dfa8cc6cc6b5ee470a711df32d25/client/db/src/lib.rs#L1940-L1949

For example, two `CommitSet` `c1` and `c2`, `apply_pending` is called for `c2` first, if later `revert_pending` need to call for `c1` there is nothing left to revert since previous `apply_pending` alreay apply all the change (not sure if this problem exist in master branch).

Adding some counter state to identify which `CommitSet` the `apply_pending/revert_pending` is calling for may workaround this problem, but it is still tricky to recover from case like some new blocks success commit to db but some older blocks fail.

Reply:
That's an issue I did not foresee :(
A third possibility would be to just not prune and prune more when available in db.
On `prune_one`, we prune from `last pruned` to `last canonicalized - range` and skip if not available (maybe it requires to maintain a new last_pruned counter in db, and maybe something else for warp synch).
@arkpar, this issue makes me wonder if there may be some unpruned content, I mean if currently we may prune over content that is still in a commit set during synch.

Reply:
> For example, two CommitSet c1 and c2, apply_pending is called for c2 first, if later revert_pending need to call for c1 there is nothing left to revert since previous apply_pending alreay apply all the change (not sure if this problem exist in master branch).

`commit_operatoin` only happens in one thread. There's a higher level lock in the `Client` that synchronises block import operations. This low level API is prone to misuse indeed. I'm considering removing the pending state alltogether and simply re-create the statedb object to recover from error.

> @arkpar, this issue makes me wonder if there may be some unpruned content, I mean if currently we may prune over content that is still in a commit set during synch.

This should not be happening. Blocks are imported one at a time and we only prune finalized blocks.

Reply:
The issue is fixed, PTAL again.